### ISSUE TITLE
Standard QGIS functions are not evaluated in docker lizmap qgis server during printing

### ISSUE THREAD

**Issue by NyakudyaA** [3liz/py-qgis-server#18]
## Procedure 3liz qgis server image
* Download the sample QGIS image attached.
* Use the 3liz docker qgis server image to serve the project. 

**Print Request**
http://localhost:8091/ows/?DPI=100&EXCEPTIONS=application/vnd.ogc.se_inimage&FORMAT=pdf&LAYERS=index1in50k,ProvincialBoundary_sae&MAP=liz/error.qgs&MAP0:EXTENT=2972996.901597,-3761252.0445,3009497.779341,-3739151.513044&MAP0:GRID_INTERVAL_X=5000&MAP0:GRID_INTERVAL_Y=5000&MAP0:LAYERS=ProvincialBoundary_sae&MAP0:SCALE=100000&MAP0:STYLES=default,default&OPACITIES=255,255&REQUEST=GetPrint&SERVICE=WMS&SRS=EPSG:3857&STYLES=default,default&TEMPLATE=map&TRANSPARENT=true&VERSION=1.3.0

Print Request Image

![print-error](https://user-images.githubusercontent.com/2510900/91570929-84031b00-e946-11ea-9afe-2c68dbf16954.png)

Image version
![vers](https://user-images.githubusercontent.com/2510900/91576988-63d45b80-e948-11ea-82a2-1757ec245fd9.png)

# Procedure camp to camp image
* Download the sample QGIS image attached.
* Use the camp to camp qgis server image to serve the project. 

```
docker run -d -p 8380:80 -v `pwd`/liz:/etc/qgisserver -e QGIS_PROJECT_FILE=/etc/qgisserver/error.qgs camptocamp/qgis-server:3.14

```
**Print Request**
http://localhost:8380/ows/?DPI=100&EXCEPTIONS=application/vnd.ogc.se_inimage&FORMAT=pdf&LAYERS=index1in50k,ProvincialBoundary_sae&MAP=/etc/qgisserver/error.qgs.qgs&MAP0:EXTENT=2972996.901597,-3761252.0445,3009497.779341,-3739151.513044&MAP0:GRID_INTERVAL_X=5000&MAP0:GRID_INTERVAL_Y=5000&MAP0:LAYERS=ProvincialBoundary_sae&MAP0:SCALE=100000&MAP0:STYLES=default,default&OPACITIES=255,255&REQUEST=GetPrint&SERVICE=WMS&SRS=EPSG:3857&STYLES=default,default&TEMPLATE=map&TRANSPARENT=true&VERSION=1.3.0

Print Request Image

![image-print](https://user-images.githubusercontent.com/2510900/91572466-e9570c00-e946-11ea-94a6-5f8f84233ab3.png)

[liz.zip](https://github.com/3liz/lizmap-web-client/files/5142487/liz.zip)

**Comment by Gustry** [comment]
> Please move the issue to the appropriate project...

There is a button for that ;-)

This is not related to Lizmap @NyakudyaA

**Comment by dmarteau** [comment]
@NyakudyaA: py-qgis-server use the official qgis debian distribution https://www.qgis.org/en/site/forusers/alldownloads.html#debian-ubuntu and is a python layer over the qgis server API. 

I don't if how it differs from the camptocamp image since I couldn't find their Dockerfile,  but as we pass all requests to the API, it may be a problem with the official distro. We need to check it out...

**Comment by NyakudyaA** [comment]
Thanks @Gustry i was printing initially from lizmap webclient then noticed this behavior. I then setup the individual containers to try to debug what was happening

**Comment by dmarteau** [comment]
@rldhont @Gustry An Idea why qgis would not evaluate function ?

**Comment by dmarteau** [comment]
@NyakudyaA Did you try a GetMap request ?

**Comment by NyakudyaA** [comment]
> @NyakudyaA Did you try a GetMap request ?

getmap request are working fine

**Comment by NyakudyaA** [comment]
> @NyakudyaA: py-qgis-server use the official qgis debian distribution https://www.qgis.org/en/site/forusers/alldownloads.html#debian-ubuntu and is a python layer over the qgis server API.
> 
> I don't know how it differs from the camptocamp image since I couldn't find their Dockerfile, but as we pass all requests to the API, it may be a problem with the official distro. We need to check it out...

You can also try to run the https://github.com/gem/oq-qgis-server. It has  a Dockerfile visible and correctly evaluates as well

**Comment by dmarteau** [comment]
> getmap request are working fine

Interesting, so its seems that expressions are not evaluated only in the context of a getprint

**Comment by dmarteau** [comment]
> You can also try to run the https://github.com/gem/oq-qgis-server. It has a Dockerfile visible and correctly evaluates as well

@NyakudyaA Thx for the tip, also there is no différence the way Qgis is installed, the only difference is how the servir is run:
- py-qgis-server: python wrapper 
- camptocamp and gem: fcgi C++ wrapper.

I  suspect a missing configuration that would prevent getprint to evaluate  function in the context of the server.

**Comment by dmarteau** [comment]
@rldhont @Gustry   I found that: if the we create the project from the QgsConfigCache  then the getPrint evaluate correctly the function but
if we create the project explicitely with QgsProject() then the getPrint does not evaluate functions.

So

```
settings = QgsServerSettings()
project = QgsConfigCache.instance().project( config_path, settings )
server.handleRequest(request, response, project=project)
```
does lead to correct  getPrint() request, while
```
project = QgsProject()
project.read( config_path )
server.handleRequest(request, response, project=project)
```
will occurs in incorrect getPrint().

Looking at the  3.10 source code  of QgsConfigCache, the cache does nothing more (except for bad layer handler): So I suspect
that the getPrint() code use the default config ache at some point. @rldhont @Gustry  any clue ?

**Comment by dmarteau** [comment]
@rldhont @Gustry I think I have found the problem:   the QgsConfigCache do `QgsProject::setInstance(...)`, I suspect that 
some internal qgis still use the `QgsProject::instance` method (seems to be removed since Qgis 3.15).   

If you instanciate the QgsProject and pass it to the server then you bypass the QgsConfigCache and the `setInstance` is never called.   Infortunatly this method is not accessible from python, so  this defeat the purpose to explicitely pass a QgsProject to the `handleRequest` method.

We need to make a PR to fix the problem so that `QgsProject::setInstance` is called whe we pass a project or eventually expose the setInstance method.

Remember that we do not use the QgsConfigCache because is does support nothing else that project files - and we want to support alternative storage (aka postgres storage)

**Comment by dmarteau** [comment]
I confirm that running a qgis server with patch applied to call `QgsProject:setInstance()` solve the problem.

**Comment by dmarteau** [comment]
@rldhont  can you call for review https://github.com/qgis/QGIS/pull/38488 ?

**Comment by dmarteau** [comment]
The fix has been accepted and backported to 3.10 and 3.14
- https://github.com/qgis/QGIS/pull/38508
- https://github.com/qgis/QGIS/pull/38509

@NyakudyaA we wait for the next qgis minor release to comfirm the fix.

**Comment by NyakudyaA** [comment]
> The fix has been accepted and backported to 3.10 and 3.14
> 
>     * [qgis/QGIS#38508](https://github.com/qgis/QGIS/pull/38508)
> 
>     * [qgis/QGIS#38509](https://github.com/qgis/QGIS/pull/38509)
> 
> 
> @NyakudyaA we wait for the next qgis minor release to confirm the fix.

Thanks a lot for spotting the bug. I will wait for the bugfix and then test

**Comment by dmarteau** [comment]
Fixed and backported to qgis 3.10
